src/synapse/dataProvider.js

```diff
@@ -202,12 +203,38 @@ const dataProvider = {
     const res = resourceMap[resource];
 
     const endpoint_url = homeserver + res.path;
-    return jsonClient(`${endpoint_url}/${params.data.id}`, {
-      method: "PUT",
-      body: JSON.stringify(params.data, filterNullValues),
-    }).then(({ json }) => ({
-      data: res.map(json),
-    }));
+
+    // In case there is a avatar_file object, save it in the media repository
+    // and update the avatar_url.
+    const f = params?.data?.avatar_file?.rawFile;
+    if (f instanceof File) {
+      const file_endpoint =
+        homeserver + "/_matrix/media/r0/upload?filename=" + f.name;
+      const headers = new Headers();
+      headers.append("Content-Type", f.type);
+      const options = {
+        method: "POST",
+        body: f,
+        headers: headers,
+      };
+      return jsonClient(file_endpoint, options).then(r => {
+        params.data.avatar_url = r.json.content_uri;
+
+        return jsonClient(`${endpoint_url}/${params.data.id}`, {
+          method: "PUT",
+          body: JSON.stringify(params.data, filterNullValues),
+        }).then(({ json }) => ({
+          data: res.map(json),
+        }));
+      });
+    } else {
```

Comment:
This makes the `update` function pretty complicated. I suggest to add another function `upload` to the DataProvider and use it directly in the UserEdit component (https://marmelab.com/react-admin/Actions.html#usedataprovider-hook). Then you can upload the file directly when it's selected and provide appropriate feedback (e.g. file is too large).